diff --git a/extdata/makeGroupFile.R b/extdata/makeGroupFile.R
index 21e99a10..6f73946f 100644
--- a/extdata/makeGroupFile.R
+++ b/extdata/makeGroupFile.R
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S pixi run --manifest-path ../pixi.toml Rscript
+#!/usr/bin/env Rscript
 
 options(stringsAsFactors = F)
 
diff --git a/extdata/step1_fitNULLGLMM_qtl.R b/extdata/step1_fitNULLGLMM_qtl.R
index 8bec4453..878083a8 100644
--- a/extdata/step1_fitNULLGLMM_qtl.R
+++ b/extdata/step1_fitNULLGLMM_qtl.R
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S pixi run --manifest-path ../pixi.toml Rscript
+#!/usr/bin/env Rscript
 
 options(stringsAsFactors = F)
 
diff --git a/extdata/step2_tests_qtl.R b/extdata/step2_tests_qtl.R
index 2ae86292..dc511c61 100644
--- a/extdata/step2_tests_qtl.R
+++ b/extdata/step2_tests_qtl.R
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S pixi run --manifest-path ../pixi.toml Rscript
+#!/usr/bin/env Rscript
 
 # options(stringsAsFactors=F, scipen = 999)
 options(stringsAsFactors = F)
diff --git a/extdata/step2b_QC_qtl.R b/extdata/step2b_QC_qtl.R
index 2ae86292..dc511c61 100644
--- a/extdata/step2b_QC_qtl.R
+++ b/extdata/step2b_QC_qtl.R
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S pixi run --manifest-path ../pixi.toml Rscript
+#!/usr/bin/env Rscript
 
 options(stringsAsFactors = F)
 suppressPackageStartupMessages({
diff --git a/extdata/step3_gene_pvalue_qtl.R b/extdata/step3_gene_pvalue_qtl.R
index d8037e70..bbe986e8 100644
--- a/extdata/step3_gene_pvalue_qtl.R
+++ b/extdata/step3_gene_pvalue_qtl.R
@@ -1,4 +1,4 @@
-#!/usr/bin/env -S pixi run --manifest-path ../pixi.toml Rscript
+#!/usr/bin/env Rscript
 
 # options(stringsAsFactors=F, scipen = 999)
 options(stringsAsFactors = F)
